### PR TITLE
App backups

### PR DESCRIPTION
**Pull request by jvanbaarsen**
This PR will introduce backups of the app you run on Intercity.
Since we assume that usergen. content is on S3 or a similar service, we only backup the "services" for now.

## Todo
* [x] Create backup infrastructure
* [x] Setup Postgres backups
* [x] Setup Mysql backups
* [x] Setup Redis backups
* [x] Setup cron for running backups
* [x] Symlink the backups directory from Intercity to Intercity-Docker

## Screenshot
**With backups enabled**
image

**Without backups enabled**
image

**App overview**
image

Fixes: #38

### CODE REVIEW

File: app/views/backups/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+          <th>Backup date</th>
+          <th>Type</th>
+          <th></th>
+        </tr>
+      </thead>
+      <tbody>
+        <%= render @backups %>
+        <tr class="<%= "hidden" if @backups.any? %>" data-role="empty-table">
+          <td colspan="3" class="text-muted text-center">There are no backups</td>
+        </tr>
```
Comment (joshuajansen) [inline app/views/backups/index.html.erb]: What about moving this empty state to a partial and doing `<%= render @backups || render "no_backups" %>`?

File: app/views/backups/status.js.erb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<% if @backup.running? %>
+    new ServerPoller("<%= status_server_app_backup_path(@backup.app.server, @backup.app, @backup) %>").poll();
```
Comment (joshuajansen) [inline app/views/backups/status.js.erb]: Indentation seems off here

File: app/views/backups/status.js.erb [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+<% else %>
+  var backup = $("#<%= dom_id(@backup) %>");
```
Comment (joshuajansen) [inline app/views/backups/status.js.erb]: No need to assign this var imo. Also a bit inconsistent use of semicolons #nitpicking

File: app/views/backups/index.html.erb (hunk 1 above)
Comment (jvanbaarsen) [inline app/views/backups/index.html.erb]: I'm doing some fancy stuff here with remote deletion. And when it gets deleted, jquery shows this TR.

File: lib/tasks/ssh_exec.rake [hunk 4]
```diff
@@ -2,19 +2,18 @@ namespace :ssh_exec do
   desc "Test interaction with SSH"
   task test: :environment do
     s = Server.first
+    output = ""
     SshExecution.new(s).execute_with_block do |ssh|
       ssh.open_channel do |ch|
-        ch.exec "dokku apps:destroy test3" do |ch, success|
+        ch.exec "dokku postgres:export formbox_postgres" do |ch, success|
```
Comment (joshuajansen) [inline lib/tasks/ssh_exec.rake]: `formbox_postgres` ?

File: app/views/backups/status.js.erb (hunk 2 above)
Comment (jvanbaarsen) [inline app/views/backups/status.js.erb]: You seem to have that right ✌️

File: app/jobs/create_backup_job.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  def backup_name
+    @backup_name ||= "#{backup.app.clean_name}-#{service.name}-#{DateTime.now.to_s.parameterize}"
+  end
+
+  def run_backup
+    SshExecution.new(app.server).execute(command: backup_command)
+    SshExecution.new(app.server).scp(from: "/root/#{backup_name}",
+                                     to: Rails.root.join("backups", "#{app.clean_name}", backup_name))
+    SshExecution.new(app.server).execute(command: "rm #{backup_name}")
```
Comment (joshuajansen) [inline app/jobs/create_backup_job.rb]: I have no knowledge of the mechanics of `SshExecution` but I assume they execute subsequently? I mean they don't fire off a command to the server and then just continue with the next?

File: app/views/backups/status.js.erb (hunk 3 above)
Comment (jvanbaarsen) [inline app/views/backups/status.js.erb]: :+1:

File: lib/tasks/ssh_exec.rake (hunk 4 above)
Comment (jvanbaarsen) [inline lib/tasks/ssh_exec.rake]: This is my test file to play with ssh tasks, I think I should add this to git ignore xD

File: app/jobs/create_backup_job.rb (hunk 5 above)
Comment (jvanbaarsen) [inline app/jobs/create_backup_job.rb]: They wait for each other, indeed

File: app/views/backups/index.html.erb (hunk 1 above)
Comment (joshuajansen) [inline app/views/backups/index.html.erb]: Ah okay, sure :) I usually just wrap it in a container and re-render the entire partial/empty state in a `js.erb` response. But this is fine too :)

**Comment by joshuajansen** [comment]
Code looks good! Also like the interface! Awesome :+1:

File: app/views/backups/index.html.erb (hunk 1 above)
Comment (jvanbaarsen) [inline app/views/backups/index.html.erb]: Not sure what you mean? As in, you render the partial in the response? (That might be a nicer solution :D)

Comment (joshuajansen) [inline app/views/backups/index.html.erb]: Jep, basically what you already do in the `status.js.erb` :smile:

Comment (jvanbaarsen) [inline app/views/backups/index.html.erb]: Awesome! Will refactor! Also, Like the new avatar!

**Pull request merged**